addons/hr_holidays/models/hr_leave.py

```diff
@@ -957,22 +953,25 @@ def action_validate(self):
                         # Could happen for part-time contract, that time off is not necessary
                         # anymore.
                         if after_leave.date_from < after_leave.date_to:
-                            split_leaves |= self.env['hr.leave'].with_context(
-                                tracking_disable=True,
-                                mail_activity_automation_skip=True,
-                                leave_fast_create=True
-                            ).create(after_leave._convert_to_write(after_leave._cache))
-                    for split_leave in split_leaves:
-                        if target_state == 'draft':
-                            continue
-                        if target_state == 'confirm':
-                            split_leave.action_confirm()
-                        elif target_state == 'validate1':
-                            split_leave.action_confirm()
-                            split_leave.action_approve()
-                        elif target_state == 'validate':
-                            split_leave.action_confirm()
-                            split_leave.action_validate()
+                            split_leaves_vals.append(after_leave._convert_to_write(after_leave._cache))
+
+                split_leaves = self.env['hr.leave'].with_context(
+                    tracking_disable=True,
+                    mail_activity_automation_skip=True,
+                    leave_fast_create=True
+                ).create(split_leaves_vals)
+
+                e_to_confirm, e_to_approve, e_to_validate = [], [], []
+                for employee, state in target_states.items():
+                    if state == 'confirm':
+                        e_to_confirm.append(employee)
+                    if state == 'validate1':
+                        e_to_approve.append(employee)
+                    if state == 'validate':
+                        e_to_validate.append(employee)
```

Comment:
Not a big concern but should be elifs. Or use a map of state -> list to append to.